File: app/src/main/java/com/loopeer/codereader/Navigator.java
package com.loopeer.codereader;

import android.app.DownloadManager;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.text.Html;
import android.widget.Toast;

import com.loopeer.codereader.coreader.db.CoReaderDbHelper;
import com.loopeer.codereader.model.Repo;
import com.loopeer.codereader.sync.DownloadRepoService;
import com.loopeer.codereader.ui.activity.AboutActivity;
import com.loopeer.codereader.ui.activity.AddRepoActivity;
import com.loopeer.codereader.ui.activity.CodeReadActivity;
import com.loopeer.codereader.ui.activity.MainActivity;
import com.loopeer.codereader.ui.activity.SearchActivity;
import com.loopeer.codereader.ui.activity.SettingActivity;
import com.loopeer.codereader.ui.activity.SimpleWebActivity;

public class Navigator {

    public final static String EXTRA_REPO = "extra_repo";
    public final static String EXTRA_ID = "extra_id";
    public final static String EXTRA_DOWNLOAD_SERVICE_TYPE = "extra_download_service_type";
    public final static String EXTRA_DIRETORY_ROOT = "extra_diretory_root";
    public final static String EXTRA_DIRETORY_ROOT_NODE_INSTANCE = "extra_diretory_root_node_instance";
    public final static String EXTRA_DIRETORY_SELECTING = "extra_diretory_selecting";
    public final static String EXTRA_WEB_URL = "extra_web_url";
    public final static String EXTRA_HTML_STRING = "extra_html_string";

    public static void startMainActivity(Context context) {
        Intent intent = new Intent(context, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
        context.startActivity(intent);
    }

    public static void startCodeReadActivity(Context context, Repo repo) {
        Intent intent = new Intent(context, CodeReadActivity.class);
        intent.putExtra(EXTRA_REPO, repo);
        context.startActivity(intent);
    }

    public static void startWebActivity(Context context, String url) {
        Intent intent = new Intent(context, SimpleWebActivity.class);
        intent.putExtra(EXTRA_WEB_URL, url);
        context.startActivity(intent);
    }

    public static void startAboutActivity(Context context) {
        Intent intent = new Intent(context, AboutActivity.class);
        context.startActivity(intent);
    }

    public static void startComposeEmail(Context context, String[] addresses, String subject, String content) {
        Intent intent = new Intent(Intent.ACTION_SENDTO, Uri.parse("mailto:"));
        intent.putExtra(Intent.EXTRA_SUBJECT, subject);
        intent.putExtra(Intent.EXTRA_EMAIL, addresses);
        intent.putExtra(Intent.EXTRA_TEXT, Html.fromHtml(content));
        if (intent.resolveActivity(context.getPackageManager()) != null) {
            context.startActivity(intent);
        } else {
            Toast.makeText(context, R.string.about_email_app_not_have, Toast.LENGTH_SHORT).show();
        }
    }

    public static void startDownloadNewRepoService(Context context, Repo repo) {
        Repo sameRepo = CoReaderDbHelper.getInstance(context).readSameRepo(repo);
        long repoId;
        if (sameRepo != null) {
            repoId = Long.parseLong(sameRepo.id);
        } else {
            repoId = CoReaderDbHelper.getInstance(context).insertRepo(repo);
        }
        repo.id = String.valueOf(repoId);
        Navigator.startDownloadRepoService(context, repo);
    }

    public static void startDownloadRepoService(Context context, Repo repo) {
        Intent intent = new Intent(context, DownloadRepoService.class);
        intent.putExtra(EXTRA_REPO, repo);
        intent.putExtra(EXTRA_DOWNLOAD_SERVICE_TYPE, DownloadRepoService.DOWNLOAD_REPO);
        context.startService(intent);
    }

    public static void startDownloadRepoService(Context context, int type) {
        Intent intent = new Intent(context, DownloadRepoService.class);
        intent.putExtra(EXTRA_DOWNLOAD_SERVICE_TYPE, type);
        context.startService(intent);
    }

    public static void startDownloadRepoServiceRemove(Context context, long downloadId) {
        Intent intent = new Intent(context, DownloadRepoService.class);
        intent.putExtra(EXTRA_DOWNLOAD_SERVICE_TYPE, DownloadRepoService.DOWNLOAD_REMOVE_DOWNLOAD);
        intent.putExtra(DownloadManager.EXTRA_DOWNLOAD_ID, downloadId);
        context.startService(intent);
    }

    public static void startSearchActivity(Context context) {
        Intent intent = new Intent(context, SearchActivity.class);
        context.startActivity(intent);
    }

    public static void startAddRepoActivity(Context context) {
        Intent intent = new Intent(context, AddRepoActivity.class);
        context.startActivity(intent);
    }

    public static void startSettingActivity(Context context) {
        Intent intent = new Intent(context, SettingActivity.class);
        context.startActivity(intent);
    }

}

File: code-reader-kt/src/main/java/com/loopeer/codereaderkt/Navigator.kt
package com.loopeer.codereaderkt

import android.app.DownloadManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.text.Html
import android.widget.Toast
import com.loopeer.codereaderkt.db.CoReaderDbHelper
import com.loopeer.codereaderkt.model.Repo
import com.loopeer.codereaderkt.sync.DownloadRepoService
import com.loopeer.codereaderkt.ui.activity.*


open class Navigator {
    companion object {
        //静态变量这么写！！！
        val EXTRA_REPO = "extra_repo"
        val EXTRA_ID = "extra_id"
        val EXTRA_DOWNLOAD_SERVICE_TYPE = "extra_download_service_type"
        val EXTRA_DIRETORY_ROOT = "extra_diretory_root"
        val EXTRA_DIRETORY_ROOT_NODE_INSTANCE = "extra_diretory_root_node_instance"
        val EXTRA_DIRETORY_SELECTING = "extra_diretory_selecting"
        val EXTRA_WEB_URL = "extra_web_url"
        val EXTRA_HTML_STRING = "extra_html_string"
    }


    public fun startMainActivity(context: Context) {
        val intent = Intent(context, MainActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
        context.startActivity(intent)
    }

    public fun startLoginActivity(context: Context) {
        val intent = Intent(context, LoginActivity::class.java)
        context.startActivity(intent)
    }

    public fun startSettingActivity(context: Context) {
        val intent = Intent(context, SettingActivity::class.java)
        context.startActivity(intent)
    }

    public fun startAddRepoActivity(context: Context) {
        val intent = Intent(context, AddRepoActivity::class.java)
        context.startActivity(intent)
    }

    public fun startAboutActivity(context: Context) {
        val intent = Intent(context, AboutActivity::class.java)
        context.startActivity(intent)
    }

    public fun startWebActivity(context: Context, url: String) {
        val intent = Intent(context, SimpleWebActivity::class.java)
        intent.putExtra(EXTRA_WEB_URL, url)
        context.startActivity(intent)
    }

    fun startComposeEmail(context: Context, addresses: Array<String>, subject: String, content: String) {
        val intent = Intent(Intent.ACTION_SENDTO, Uri.parse("mailto:"))
        intent.putExtra(Intent.EXTRA_SUBJECT, subject)
        intent.putExtra(Intent.EXTRA_EMAIL, addresses)
        intent.putExtra(Intent.EXTRA_TEXT, Html.fromHtml(content))
        if (intent.resolveActivity(context.packageManager) != null) {
            context.startActivity(intent)
        } else {
            Toast.makeText(context, R.string.about_email_app_not_have, Toast.LENGTH_SHORT).show()
        }
    }

    fun startDownloadRepoService(context: Context, repo: Repo) {
        val intent = Intent(context, DownloadRepoService::class.java)
        intent.putExtra(EXTRA_REPO, repo)
        intent.putExtra(EXTRA_DOWNLOAD_SERVICE_TYPE, DownloadRepoService.DOWNLOAD_REPO)
        context.startService(intent)
    }

    fun startDownloadRepoService(context: Context, type: Int) {
        val intent = Intent(context, DownloadRepoService::class.java)
        intent.putExtra(EXTRA_DOWNLOAD_SERVICE_TYPE, type)
        context.startService(intent)
    }

    fun startDownloadRepoServiceRemove(context: Context, downloadId: Long) {
        val intent = Intent(context, DownloadRepoService::class.java)
        intent.putExtra(EXTRA_DOWNLOAD_SERVICE_TYPE, DownloadRepoService.DOWNLOAD_REMOVE_DOWNLOAD)
        intent.putExtra(DownloadManager.EXTRA_DOWNLOAD_ID, downloadId)
        context.startService(intent)
    }

    fun startDownloadNewRepoService(context: Context, repo: Repo) {
        val sameRepo = CoReaderDbHelper.getInstance(context).readSameRepo(repo)
        val repoId: Long
        if (sameRepo != null) {
            repoId = java.lang.Long.parseLong(sameRepo.id)
        } else {
            repoId = CoReaderDbHelper.getInstance(context).insertRepo(repo)
        }
        repo.id = repoId.toString()
        Navigator().startDownloadRepoService(context, repo)
    }

}